x-pack/plugins/security_solution/server/lib/detection_engine/rules/types.ts

```diff
@@ -329,16 +331,18 @@ export interface ReadRuleOptions {
 
 export interface DeleteRuleOptions {
   alertsClient: AlertsClient;
-  id: IdOrUndefined;
-  ruleId: RuleIdOrUndefined;
+  savedObjectsClient: SavedObjectsClientContract;
+  ruleStatusClient: RuleStatusSavedObjectsClient;
+  ruleStatuses: SavedObjectsFindResponse<IRuleStatusSOAttributes, unknown>;
+  id: Id;
 }
 
 export interface FindRuleOptions {
   alertsClient: AlertsClient;
-  perPage: PerPageOrUndefined;
-  page: PageOrUndefined;
-  sortField: SortFieldOrUndefined;
-  filter: QueryFilterOrUndefined;
-  fields: FieldsOrUndefined;
-  sortOrder: SortOrderOrUndefined;
+  perPage?: PerPageOrUndefined;
+  page?: PageOrUndefined;
+  sortField?: SortFieldOrUndefined;
+  filter?: QueryFilterOrUndefined;
+  fields?: FieldsOrUndefined;
+  sortOrder?: SortOrderOrUndefined;
```

Comment:
I would strongly advise against making everything optional compared to the way it is which is forcing people to push down `undefined`. Reason being, we have seen a lot of bugs where people get around fixing unit tests by making properties optional. Later we end up with bad bugs where people aren't sending down the required parameters.

By forcing people to push down `undefined` we force them to see each parameter to handle each one with either an `undefined` or a value when we use named parameters.

If you really feel strongly about it, I would remove the `OrUndefined` and just use `perPage?: PerPage`, but in the past to fix bugs we actually _remove_ the `?` on parameters and then manually handle all the missing locations.

Another anti-pattern is when people add a new parameter to something and then mark it with `?` to get around having to handle it everywhere, including tests and just handle it in 1 spot. That can lead to bugs as well. Although it's hard to police this type of thing, usually I try to force people to handle everything even if it looks a bit ugly than make things optional for as long as I can.

On the flip side, I undo the optional parameters in a lot of cases and then use the type checker to force me to handle all the use cases to fix the bugs in other pull requests.

Reply:
Ok, I don't have strong preferences on `?` vs. `undefined`. So if it is an established practice in the team, I'll stick to `undefined`.